Mathlib/Data/Option/Basic.lean

```diff
@@ -3,28 +3,368 @@ Copyright (c) 2017 Mario Carneiro. All rights reserved.
+
+@[simp]
+theorem guard_eq_some' {p : Prop} [Decidable p] (u) : _root_.guard p = some u ↔ p := by
+  cases u
+  by_cases p <;> simp [_root_.guard, h] <;> first |rfl|contradiction
+
+theorem lift_or_get_choice {f : α → α → α} (h : ∀ a b, f a b = a ∨ f a b = b) :
+    ∀ o₁ o₂, liftOrGet f o₁ o₂ = o₁ ∨ liftOrGet f o₁ o₂ = o₂
+  | none, none => Or.inl rfl
+  | some a, none => Or.inl rfl
+  | none, some b => Or.inr rfl
+  | some a, some b => by simpa [lift_or_get] using h a b
+
+-- @[simp]
+theorem lift_or_get_none_left {f} {b : Option α} : liftOrGet f none b = b := by cases b <;> rfl
+
+-- @[simp]
+theorem lift_or_get_none_right {f} {a : Option α} : liftOrGet f a none = a := by cases a <;> rfl
+
+@[simp]
+theorem lift_or_get_some_some {f} {a b : α} : liftOrGet f (some a) (some b) = f a b :=
+  rfl
+
+/-- Given an element of `a : option α`, a default element `b : β` and a function `α → β`, apply this
+function to `a` if it comes from `α`, and return `b` otherwise. -/
+def casesOn' : Option α → β → (α → β) → β
+  | none, n, _ => n
+  | some a, _, s => s a
+
+@[simp]
+theorem cases_on'_none (x : β) (f : α → β) : casesOn' none x f = x :=
+  rfl
+
+@[simp]
+theorem cases_on'_some (x : β) (f : α → β) (a : α) : casesOn' (some a) x f = f a :=
+  rfl
+
+@[simp]
+theorem cases_on'_coe (x : β) (f : α → β) (a : α) : casesOn' (a : Option α) x f = f a :=
+  rfl
+
+-- @[simp]
+theorem cases_on'_none_coe (f : Option α → β) (o : Option α) :
+    casesOn' o (f none) (f ∘ (fun a => ↑a)) = f o := by cases o <;> rfl
+
+theorem orelse_eq_some (o o' : Option α) (x : α) :
+    (o <|> o') = some x ↔ o = some x ∨ o = none ∧ o' = some x := by
+  cases o
+  · simp only [true_and, false_or, eq_self_iff_true, none_orelse, iff_self]
+  · simp only [some_orelse, or_false, false_and, iff_self]
+
+
+theorem orelse_eq_some' (o o' : Option α) (x : α) :
+    o.orElse (fun _ => o') = some x ↔ o = some x ∨ o = none ∧ o' = some x :=
+  Option.orelse_eq_some o o' x
+
+@[simp]
+theorem orelse_eq_none (o o' : Option α) : (o <|> o') = none ↔ o = none ∧ o' = none := by
+  cases o
+  · simp only [true_and, none_orelse, eq_self_iff_true, iff_self]
+  · simp only [some_orelse, false_and, iff_self]
+
+@[simp]
+theorem orelse_eq_none' (o o' : Option α) : o.orElse (fun _ => o') = none ↔ o = none ∧ o' = none :=
+  Option.orelse_eq_none o o'
+
+section
+
+open Classical
+
+theorem choice_eq_none (α : Type _) [IsEmpty α] : choice α = none :=
+  dif_neg (not_nonempty_iff_imp_false.mpr isEmptyElim)
+
+theorem choice_is_some_iff_nonempty {α : Type _} : (choice α).isSome ↔ Nonempty α := by
+  constructor
+  · intro h
+    exact ⟨Option.get _ h⟩
+  · intro h
+    simp only [choice]
+    rw [dif_pos h]
+    exact is_some_some
 
--- theorem join_eq_none {o : Option (Option α)} : o.join = none ↔ o = none ∨ o = some none := by
---   rcases o with _|_|_; simp
+end
 
-@[simp] theorem lift_or_get_none_left {f} {b : Option α} : lift_or_get f none b = b := by
-  cases b <;> rfl
+#align option.elim Option.elimₓ
```

Comment:
The `elimₓ` should be deleted, I called it `elim'` on #504 but maybe it's better to just adapt to the new argument order of `Option.elim`

Reply:
The meaning of the `ₓ` has changed slightly in newer versions of mathport. Now it means a name character which is stripped only for printing purposes, so it is used whenever an alignment is technically incorrect but will still elaborate as intended if you just find/replace uses of `Option.elimₓ` with `Option.elim`. So this pattern will basically automate the adaptation to the new argument order. (Alternatively, you can backport change the argument order of `option.elim` and make this an actual defeq match.)

Reply:
Oh wow, ok, thanks for the correction, that's neat that it automates to the new argument order!